Updates and an example for simple clustering configurations

umpc opened the pull request:
Hi,

Thank you for reading my feedback!

I took another look at if `rm -r /var/lib/foundationdb/data/*` is required and you are right that it is not, so I left it out as it's not needed.

Everything we have discussed so far works well.

These changes keep the existing behavior intact and address linking new servers into the default cluster that is created. This is actually made easier with `rm -r /var/lib/foundationdb/data/*` gone.

I removed the `sed` command setting a `0.0.0.0` listening address because during the linking of new servers, when the `fdb.cluster` file is copied the way it explains in [the documentation](https://apple.github.io/foundationdb/administration.html#adding-machines-to-a-cluster), the receiving server doesn't have the existing server's address, as it's set to `127.0.0.1`, and in the documentation, it lists under the example config file that the port is the value of the `$ID` config variable that resolves to the server ID listed in the config file, taken from the initial `fdb.cluster` file.

These changes improve the behavior of the containers to what users would likely expect while reading the existing documentation.

The new change grabs the container's IP at runtime, replaces it in the `fdb.cluster` file within the `*.default` directory, and then that file is added to volumes if it doesn't exist, as usual. This works because the best practices in using containers has them setup with only one IP address. This makes a simple `hostname` call to grab the ID of the container, searches `/etc/hosts` for the matching line, and then uses `awk` to select the first space-separated value, which is the container's egress IP.

I also added an example of a simulated fault-tolerance setup to the readme, so that anyone can become familiar with having a replicated and fault-tolerant cluster within a few minutes.

I understand that the changes might could be improved or you may not agree, so I am opening this first here for discussion before porting this to the other OSes. Currently, this works perfectly on Ubuntu 18.04.

I looked at how other projects that have multiple base images have set up their versioning and added the foundationdb semantic version to the build tag, with where you had the OS and version originally. Because there isn't exactly an explicit default base image OS with Docker images, I changed instances of `latest` to `5.1.7-ubuntu-18.04` in the readme, so that users know immediately that they have a choice, and know which base system they are using to run FoundationDB.

Please let me know your thoughts. Thanks for your work on this!

Edit: If you are okay with these changes, do you think that the `CMD` instruction should be made into its own script that each OS's Dockerfile can reference to cut down on repeated code, or do you think that could add more work for maintainers?

umpc commented:
I found that since `--init` was dropped, `bash` doesn't automatically forward signals to child processes. Also, arguments could not be passed to `fdbmonitor`, such as `--lockfile` or `--loggroup`.

The latest commit fixes both of those issues.

umpc commented:
Cluster ID collisions were possible because it was only set during the image build, so I added a check to fix it during startup when `fdb.cluster` doesn't exist.

I also made it delete the original log files created during build from the auto-start.

bjrnio commented:
1. Haven't tested it but IPv6 addresses should work seamlessly. Pretty nice clean way of getting the container's IP.
2. I think your point about letting the users know about version choice is right, let's keep it like that.
3. Regarding scripts on `ENTRYPOINT`s and `CMD`s, my opinion is 100% opposing, it makes it harder for reviewers, makes it look less transparent and feels like a fix for a problem that is not that big of a deal, which is the prettiness of the Dockerfile. It's what Dockerfiles of most products do but I do not encourage that practice, feels _hacky_.
That said, if the script contains logic, such as the one you added, then we can talk about it. I personally would put it in the Dockerfile as it's not that long, but I understand that once there is logic involved, syntax and readability may get priority.

bjrnio commented:
Something like this (I've done it in 1min so it's surely wrong, but I mean the concept), putting everything in the Dockerfile as long as it doesn't get too large, I feel it's OK.

```Dockerfile
CMD /bin/sh -c "\
    if [ ! -f /etc/foundationdb/fdb.cluster ]; then
      sed -i s/^@/\"$(mktemp -u XXXXXXXX)\"@/ /etc/foundationdb.default/fdb.cluster
      sed -i s/@.*:/@\"$(grep $(hostname) /etc/hosts | awk '{print $1}')\":/ /etc/foundationdb.default/fdb.cluster
    fi
    for DIR in $FDB_USER_DIRS; do find $DIR.default -maxdepth 1 -exec cp -r --no-clobber {} $DIR \;; done
    if [ \"$(id -u foundationdb)\" != $FDB_UID -o \"$(id -g foundationdb)\" != $FDB_GID ]; then
      groupmod -g $FDB_GID --non-unique foundationdb
      usermod -g $FDB_GID -u $FDB_UID --non-unique foundationdb
      chown -R foundationdb:foundationdb $FDB_USER_DIRS
    fi"
    ...more
```

It's cryptic, but I feel easier to understand too, because it's right there in the same file that all other image configuration is.

However, I can understand if you decide to put it elsewhere. Do however you feel it's right, I believe there is no written guidelines for this matter.

umpc commented:
I agree. I'm not a fan of using a script ether.

I would be fine with switching back as long as `fdbmonitor` is somehow shutdown smoothly, there's no leftover log files from build, and we can get confirmation that running `chown` isn't going to slow startup times after a long time from a large number of files being created, even though I doubt that this would be an issue in practice. The Dockerfile also then wouldn't support passing arguments in, though this isn't a huge issue with `fdbmonitor`.

The current script fixes these issues and will prevent cluster ID collisions once the image is posted to Docker Hub. I am not sure how `fdbmonitor` shutdown should be handled without a script though. If we do go with a script, I just don't want it to end up being as large and complex as the [`postgres` `docker-entrypoint.sh`](https://github.com/docker-library/postgres/blob/master/10/docker-entrypoint.sh)

bjrnio commented:
Let's go with Dockerfile-only then.

`chown`'s performance I believe has no solution and `fdbmonitor` is not going to get any parameters out-of-the-box so I think we are good!

umpc commented:
I was saying that to essentially ask if you had any ideas for how that might work because I don’t think fixing the first issue I mentioned is reasonably possible in only a Dockerfile without creating more of a mess. 

I would prefer not to use separate scripts, though that seems to be the best option so far.

bjrnio commented:
> I was saying that to essentially ask if you had any ideas for how that might work because I don’t think fixing the first issue I mentioned is reasonably possible in only a Dockerfile without creating more of a mess.

I believe there is no issue with long strings as long as some format tweaking is put in place, it's not clean but calling external resources is just as dirty.


> Another observation is that the way we are syncing UID/GIDs might be incompatible with Windows. 

At that point wouldn't it be better to use [microsoft/windowsservercore](https://hub.docker.com/r/microsoft/windowsservercore/) and install FoundationDB for Windows?

umpc commented:
That is another design problem with Docker where I don’t really understand their perspective.

They create a build context for each Dockerfile directory, so unless we want to make another script to manage copying and building, each directory requires its own entrypoint script.

For example, using anything in ../ relative to the build directory won’t work. At this point, I would just prefer copying the entrypoint scripts rather than another level of abstraction that could be error-prone or misunderstood.

Also, the scripts cannot be only 4 lines with the same functionality because they must handle adding the container IP, setting the correct permissions, ensuring a unique cluster ID is made for new installs, and most importantly, trap and wait for signals to properly shut down fdbmonitor instead of sending it a SIGKILL.

It is just not an ideal situation though I think that we are making the best of it. I would just like more feedback from others, as there are probably edge cases that should get more attention.

I am a bit busy at the moment though will get back to this soon.

umpc commented:
I have switched from using bind mounts to volumes for the database state. This makes the Dockerfile work with `fsync` on macOS if it wasn't working already. I tested the steps in the readme and it worked fine on macOS 10.13.

I only have Windows 10 available in a VM, which cannot run Docker, so it would be excellent if someone running Windows on bare metal could test this and report back. It likely now works normally on Windows.

I have updated the readme accordingly and added a few grammatical changes. I will apply these changes to the other systems and then this PR specifically is ready to be merged.

umpc commented:
All finished. I know that you wanted CentOS 7, so I added it. I will ask on my next reply to the main PR whether the core devs are open to supporting all of the extra OSes, just the ones that they already support, or just one base image. I will also ask them if there is anything that they want added as far as pushing to Docker Hub.

bjrnio commented on docker/centos/7.4/Dockerfile:
```diff
@@ -0,0 +1,32 @@
+ARG FDB_VERSION="5.1.7"
+ARG FDB_RPM_RELEASE="1"
+
+ARG FDB_PKG_URL="https://www.foundationdb.org/downloads/${FDB_VERSION}/rhel6/installers"
```
I want CentOS 7 but you are installing RHEL/CentOS 6 packages, whilst there is a RHEL/CentOS 7 package available [here](https://www.foundationdb.org/downloads/5.1.7/rhel7/installers/foundationdb-server-5.1.7-1.el7.x86_64.rpm).

If it works, keep it, but I had some issues with systemd in CentOS 7 with the aforementioned package.

umpc replied:
Nice catch. Do you know if they say that they support rhel7 officially somewhere? (Other than the packaging scripts and download URL, of course.)

bjrnio replied:
They do provide installation instructions [here](https://apple.github.io/foundationdb/getting-started-linux.html?highlight=RHEL/CentOS).
Looks like they do provide support for RHEL 7 but forgot to say so in www.foundationdb.org/download

bjrnio commented:
And regarding other OSs, i think it's not just adding support. IMHO the build process is a mess, where did the old and universal `./configure && make && make install` go?
And as long as the build process remains a mess, it'll be hard to add support for anything else than the niche platforms they offer.

bjrnio commented:
Anyways, great job!

bjrnio commented on docker/ubuntu/18.04/entrypoint.sh:
```diff
@@ -0,0 +1,30 @@
+then
+	groupmod -g $FDB_GID --non-unique foundationdb
+	usermod -g $FDB_GID -u $FDB_UID --non-unique foundationdb
+	chown -R foundationdb:foundationdb $FDB_USER_DIRS
```
If some of the directories in `FDB_USER_DIRS` do not exist this will kill execution.
So if a user just wants the simplest of the FountationDB instances and runs without any mounts, this will break:
```
$ cd docker/ubuntu/18.04

$ docker build -t foundationdb:5.1.7-ubuntu-18.04 .
  ...
  Successfully built 4c4b9fca1e02
  Successfully tagged foundationdb:5.1.7-ubuntu-18.04

$ docker run -it -p 4500:4500 foundationdb:5.1.7-ubuntu-18.04
chown: cannot access '/var/log/foundationdb': No such file or directory

bjrnio replied:
I guess this happens on all `entrypoint.sh` files, too

bjrnio commented on docker/ubuntu/18.04/entrypoint.sh:
```diff
@@ -0,0 +1,30 @@
+then
+	# Replace the default IP address with the container's IP.
+	CONTAINER_IP=$(grep $(hostname) /etc/hosts | awk '{print $1}')
+	sed -i s/@.*:/@$CONTAINER_IP:/ /etc/foundationdb.default/fdb.cluster
```
This leaves the `fdb.cluster` file with the container's IP, which is great if you want to create a cluster, but not if I'm running an instance of FoundationDB and connecting from local `fdbcli` or a local API.

Copying the file to a non-volume location, replacing the container's IP with `127.0.0.1` and having bound the container's 4500/tcp port to my local machine's, doesn't work either.

```
$ cp etc/fdb.cluster .
$ vi fdb.cluster  # Replaced 172.17.0.2 with 127.0.0.1
$ fdbcli
Using cluster file `fdb.cluster'.

The database is unavailable; type `status' for more information.

Welcome to the fdbcli. For help, type `help'.
fdb> status

Using cluster file `fdb.cluster'.

Unable to communicate with the cluster controller at 172.17.0.2:4500 to get
status.

Configuration:
  Redundancy mode        - unknown
  Storage engine         - unknown
  Coordinators           - unknown

Cluster:
  FoundationDB processes - unknown
  Machines               - 
  Machines               - unknown

Data:
  Replication health     - unknown
  Moving data            - unknown
  Sum of key-value sizes - unknown
  Disk space used        - unknown

Operating space:
  Unable to retrieve operating space status

Workload:
  Read rate              - unknown
  Write rate             - unknown
  Transactions started   - unknown
  Transactions committed - unknown
  Conflict rate          - unknown

Backup and DR:
  Running backups        - 0
  Running DRs            - 0

Client time: 05/19/18 10:41:17

fdb> 
```

bjrnio replied:
Am I wrong or does this issue persist? Have you tested it?
If I build ubuntu-18.04 (haven't tested the others) and run it with the instructions from README.md, I can't connect from `fdbcli` using the given fdb.cluster file from the container.

bjrnio commented on docker/README.md:
```diff
@@ -51,18 +54,163 @@ version: '3'
 
-  db:
-    image: foundationdb:latest
+  fdb:
```
I'm not going to push very hard on this one, but `fdb` is not a good name for a service. Either `db` or `database`,  but specifying the product in the service name makes it harder to understand a stack you are not familiar with or that you didn't write.

umpc replied:
I will change this back as I am less familiar with some of the use-cases of Compose.

I was looking to avoid a potential naming collision in cases where FoundationDB is not the only database used in a group.

umpc commented:
> And regarding other OSs, i think it's not just adding support. IMHO the build process is a mess, where did the old and universal `./configure && make && make install` go?

Ah, yes. I too hope that is cleaned up at some point soon.

> And as long as the build process remains a mess, it'll be hard to add support for anything else than the niche platforms they offer.

I was getting more at how any OS that Docker supports could use any of those distros with the usage appearing to be the same no matter which base image is used, meaning a minimalistic choice of only the latest Ubuntu LTS, I imagine, would still work for most people.

You have found some excellent points to address. I can't respond to all of them today, though I will soon.

bjrnio commented:
No issue whatsoever, I understand that you are busy right now and greatly appreciate that you are finding time windows here and there!

> I was getting more at how any OS that Docker supports could use any of those distros with the usage appearing to be the same no matter which base image is used, meaning a minimalistic choice of only the latest Ubuntu LTS, I imagine, would still work for most people.

Yeah probably, but `16.04` and `6.9` are still in their support periods, and as you pointed out, it's almost the same no matter what OS layer you use, so it's not hard for us to maintain and provides alternative choice to those that, for whatever reason, need an older OS release.
For example, say you want to use Ubuntu, but we all know Canonical and it's x.0 releases, so you'd rather use 16.04 by now, and upgrade down the path.

umpc commented:
I've made the changes you have requested.

umpc commented:
Were you able to fix what you were having trouble with? I saw a comment in email but not here.

bjrnio commented:
No, the thing is, if I follow the README's instructions, I get a FDB server in a container with 4500/tcp bound to my host's 4500/tcp, and a fdb.cluster file, but I can't connect to the database using a local `fdbcli` and the aforementioned cluster file.

Any ideas? Could it be related with FDB's listen address?

Note: I don't really have the time right now so I can't work on this as much.

umpc commented:
I got around to testing this and the README is fine.

Here are the exact steps that I took, closely following the README:

```bash
git clone https://github.com/umpc/foundationdb
cd foundationdb/
git checkout docker
cd ..
docker build -t foundationdb:5.1.7-ubuntu-18.04 foundationdb/docker/ubuntu/18.04
mkdir fdbtest
cd fdbtest/
mkdir -p {etc,log}
docker run -d \
  -e FDB_UID=$(id -u) \
  -e FDB_GID=$(id -g) \
  --mount type=volume,src=fdb-example-data,dst=/var/lib/foundationdb \
  --mount type=bind,src=$(pwd)/etc,dst=/etc/foundationdb \
  --mount type=bind,src=$(pwd)/log,dst=/var/log/foundationdb \
  --name fdb-example \
  -p 127.0.0.1:4500:4500 \
  foundationdb:5.1.7-ubuntu-18.04
fdbcli -C etc/fdb.cluster --exec "status"
```

After typing out this comment, I ran ```fdbcli -C etc/fdb.cluster --exec "status"``` again and the single-node cluster was "Healthy."

umpc commented:
The FDB team is taking care of this now.
